lib/handler.js

```diff
@@ -96,7 +97,12 @@ internals.handler = function (request, callback) {
 
     // Execute handler
 
-    request.route.settings.handler.call(bind, request, reply);
+    const handlerResult = request.route.settings.handler.call(bind, request, reply);
+
+    // Give hapi a fighting chance to deal with uncaught rejections.
+    if (Promises.isThennable(handlerResult)) {
+        handlerResult.catch(reply);
```

Comment:
It also needs to handle rejections where the object is not an `instanceof Error`, like this code:
https://github.com/hapijs/hapi/blob/b4aeaf77cf3f74fa12b2a514c01748768d974e4a/lib/response.js#L471-L480.